engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala

```diff
@@ -27,34 +27,71 @@ trait FlinkSource extends Source {
   * @tparam Raw - type of raw event that is generated by flink source function.
   *             This is needed to handle e.g. syntax suggestions in UI (in sources with explicite @MethodToInvoke).
   */
-trait FlinkSourceTestSupport[Raw] extends SourceTestSupport[Raw] { self: Source with FlinkIntermediateRawSource[Raw] =>
+trait FlinkSourceTestSupport[Raw] extends SourceTestSupport[Raw] with ExplicitTypeInformationSource[Raw] {
+  self: FlinkSource =>
 
   // TODO: design better way of handling test data in generic FlinkSource
   // Probably we *still* want to use CollectionSource (and have some custom logic in parser if needed), but timestamps
   // have to be handled here for now
   def timestampAssignerForTest: Option[TimestampWatermarkHandler[Raw]]
 
-  def typeInformation: TypeInformation[Raw]
-
 }
 
 /**
-  * Typical source with methods specific for Flink, user has only to define Source function, type information and timestamp assigner.
-  * Optionally [[FlinkSourceTestSupport]] can be mixed in to provide generating test data while authoring scenarios.
-  *
-  * @tparam Raw - type of raw event that is generated by flink source function.
-  *             This is needed to handle e.g. syntax suggestions in UI (in sources with explicite @MethodToInvoke).
-  */
-trait BasicFlinkSource[Raw] extends FlinkSource with FlinkIntermediateRawSource[Raw] {
+ * Source providing standard logic for transforming a `DataStreamSource[Raw]` into `DataStream[Context]` using
+ * customizable [[ContextInitializer]] and [[TimestampWatermarkHandler]].
+ *
+ * @tparam Raw - type of raw event that is generated by flink source function.
+ */
+trait StandardFlinkSource[Raw]
+    extends FlinkSource
+    with CustomizableContextInitializerSource[Raw]
+    with CustomizableTimestampWatermarkHandlerSource[Raw]
+    with ExplicitUidInOperatorsSupport {
 
-  @silent("deprecated")
-  def flinkSourceFunction: SourceFunction[Raw]
+  def sourceStream(
```

Comment:
Can it be protected?

Reply:
Good point, haven't thought of that. Added protected and it compiles.